### PR TITLE
CDK-615: Permissions issue writing to a partitioned Hive external dataset

### PR DESCRIPTION
**Pull request by joey**
* Added a test of `load()` methods of `DatasetWriterCacheLoader` and
  `IncrementalDatasetWriterCacheLoader`.
* Moved the `writer.initialize()` call to happen before the
  `listener.partitionAdded()` call in `DatasetWriterCacheLoader#load`
  and `IncrementalDatasetWriterCacheLoader#load`.

### CODE REVIEW

**Comment by joey** [comment]
I've been thinking more about this and my fix here may not be correct. This works for file system datasets and for external Hive datasets, but for managed Hive datasets the directories should probably be created by Hive.

I'm thinking the correct solution is that the partition listener should always be called before the directory is created. The external Hive metadata provider would create it on the Kite side and the managed one would just let the Hive metastore create it.

Thoughts?

**Comment by joey** [comment]
I pushed a new commit to the branch that reverses the logic and makes partition listeners responsible for deciding if Kite or an external system should create the partition directories. This makes Kite work more consistently with both managed and external Hive tables and fixes the original permissions issue.

There is probably still follow-up work on how to deal with the permissions and ownership of files themselves. I think as it stands, you can't drop a Hive managed dataset using the Hive DDL statements as the files are probably not owned by Hive but by the user that ran Kite.

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/PartitionedDatasetWriter.java [hunk 1]
```diff
@@ -256,6 +262,9 @@ public IncrementalDatasetWriterCacheLoader(FileSystemView<E> view) {
             dataset.getNamespace(), dataset.getName(), partition.toString());
       }
 
+      // initialize the writer after calling the listener
+      // this lets the listner decide if and how to create the
```
Comment (rdblue) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/PartitionedDatasetWriter.java]: Typo: listner

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/PartitionedDatasetWriter.java [hunk 2]
```diff
@@ -216,6 +218,9 @@ public DatasetWriterCacheLoader(FileSystemView<E> view) {
...
```
Comment (rdblue) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/PartitionedDatasetWriter.java]: Typo: Listner

**Comment by rdblue** [comment]
+1

**Pull request merged**